src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/RestoreCommandHandler.cs

```diff
@@ -0,0 +1,106 @@
+        {
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+        }
+
+        public void RunSolutionRestore()
+        {
+            _joinableTaskFactory.RunAsync(async () =>
+            {
+                var cancellationTokenSource = new CancellationTokenSource();
+                _cancelBuildToken = cancellationTokenSource;
+
+                try
+                {
+                    await GetAndActivatePackageManagerOutputWindowAsync();
+
+                    await TaskScheduler.Default;
+                    IBrokeredServiceContainer serviceContainer = await _asyncServiceProvider.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+                    IServiceBroker serviceBroker = serviceContainer.GetFullAccessServiceBroker();
+
+                    var nugetSolutionService = await serviceBroker.GetProxyAsync<INuGetSolutionService>(NuGetServices.NuGetSolutionService);
+
+                    try
+                    {
+                        await nugetSolutionService.RestoreSolutionAsync(cancellationTokenSource.Token);
+                    }
+                    finally
+                    {
+                        (nugetSolutionService as IDisposable)?.Dispose();
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Only log to the activity log for now
+                    ActivityLog.LogError("NuGet Package Manager", e.Message);
+                }
+                finally
+                {
+                    cancellationTokenSource.Dispose();
+                    _cancelBuildToken = null;
+                }
+            });
+        }
+
+        public bool IsRestoreActionInProgress()
+        {
+            return _cancelBuildToken != null;
+        }
+
+        public async Task GetAndActivatePackageManagerOutputWindowAsync()
+        {
+            var outputWindow = await _asyncServiceProvider.GetServiceAsync<SVsOutputWindow, IVsOutputWindow>();
+            var vsUIShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>();
+
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            // create the Package Manager pane within the Output window
+            var hr = outputWindow.CreatePane(
+                ref GuidList.GuidNuGetOutputWindowPaneGuid,
+                Resources.OutputConsolePaneName,
+                fInitVisible: 1,
+                fClearWithSolution: 0);
+            ErrorHandler.ThrowOnFailure(hr);
+
+            IVsOutputWindowPane pane;
+            hr = outputWindow.GetPane(
+                ref GuidList.GuidNuGetOutputWindowPaneGuid,
+                out pane);
+            ErrorHandler.ThrowOnFailure(hr);
```

Comment:
Hey, this looks familiar! 😀
Have you tested how a thrown failure here gets materialized to the user? And/or if it gets logged in a Fire&Forget telemetry?

Reply:
Since this is a new method, we know all invocations are going to be in the PR's diff. We can see that the method is called in a try-catch block, and the catch explicitly says it's only being logged to the activity log currently.

Reply:
I used the activity log on purpose!
I'll create a follow up to use FileAndForget...for the time being it's actually easier for diagnostics if failures are in the activity log.

Reply:
Created a follow up https://github.com/NuGet/Home/issues/9352.